### PR TITLE
Fix spaces in uri 

### PR DESCRIPTION
**Pull request by AlekhyaYalla**
We have one repo with the registry URL having white spaces. So while trying to [read data](https://github.com/dependabot/dependabot-core/blob/0bfb0075fe547b3f4036f8ec161dee0d7d8ed254/npm_and_yarn/lib/dependabot/npm_and_yarn/update_checker/registry_finder.rb#L56) from that uri, parsing is failing due to white spaces. 
So replacing the white spaces before passing it to the get call.

### CODE REVIEW

**Comment by jurre** [comment]
@AlekhyaYalla could you add a test to verify this keeps working as intended in the future?

**Comment by jurre** [comment]
It also seems like this breaks some existing tests @AlekhyaYalla, would you mind addressing those as well?

**Comment by jeffwidman** [comment]
> We have one repo with the registry URL having white spaces

Can you share the URL (with anything private redacted)? I'm curious how you ended up with a registry URL containing white spaces?

**Comment by AlekhyaYalla** [comment]
> > We have one repo with the registry URL having white spaces
> 
> Can you share the URL (with anything private redacted)? I'm curious how you ended up with a registry URL containing white spaces?

Here is the url: https://pkgs.dev.azure.com/org/ocp ds tools/_packaging/ocp_ds_tools_PublicPackages/npm/registry/ 
When this url is being used for loading the registry, it is replacing the white spaces with "%20".

**Comment by AlekhyaYalla** [comment]
@jurre I could add unit tests for the change and also all the test cases were passing. Could you please review this change? Thanks!

File: npm_and_yarn/lib/dependabot/npm_and_yarn/update_checker/registry_finder.rb [hunk 1]
```diff
@@ -134,10 +135,13 @@ def npmrc_registries
...
             next if Regexp.last_match[:registry].include?("${")
 
+            registry = Regexp.last_match[:registry]
+            token = Regexp.last_match[:token]&.strip
+
             registries << {
               "type" => "npm_registry",
-              "registry" => Regexp.last_match[:registry],
-              "token" => Regexp.last_match[:token]&.strip
+              "registry" => registry.gsub(/\s+/, "%20"),
```
Comment (jeffwidman) [inline npm_and_yarn/lib/dependabot/npm_and_yarn/update_checker/registry_finder.rb:143]: Note to myself when I have time to review: should the `gsub` happen here or up in line 138? Compare how token does the strip up above such that down here it's just `"token" => token`... but that may affect line 154... anyway, I need to pull up the full code, not just the snippets in the web UI to see what's happening here...

File: npm_and_yarn/lib/dependabot/npm_and_yarn/update_checker/registry_finder.rb [hunk 2]
```diff
@@ -168,7 +173,8 @@ def yarnrc_registries
 
             registry = Regexp.last_match[:registry].strip.
                        sub(%r{/+$}, "").
-                       sub(%r{^.*?//}, "")
+                       sub(%r{^.*?//}, "").
+                       gsub(/\s+/, "%20")
```
Comment (jeffwidman) [inline npm_and_yarn/lib/dependabot/npm_and_yarn/update_checker/registry_finder.rb:177]: Note to myself: check into ruby perf of chaining `sub` calls like this... how are these implemented under the covers? Ie, are we re-looping all the time such that it'd be better to combine these calls? Need to research how Ruby handles this...

File: npm_and_yarn/spec/fixtures/projects/npm6/npmrc_auth_token_with_space/package.json [hunk 3]
```diff
@@ -0,0 +1,25 @@
+{
+    "name": "{{ name }}",
+    "version": "1.0.0",
+    "description": "",
+    "main": "index.js",
+    "scripts": {
+        "test": "echo \"Error: no\\ test\ specified\" && exit 1",
```
Comment (jeffwidman) [inline npm_and_yarn/spec/fixtures/projects/npm6/npmrc_auth_token_with_space/package.json:7]: note to myself: Why is github formatter showing this in red? Is it a bug in syntax highligher or a funky character not apparent to normal human eyes? Doublecheck the UTF character byte code...

**Review by jeffwidman — COMMENTED** [review]
Thanks!

I took a quick pass and looks straightforward, had a few questions that I need to dig deeper on before I can speak articulately about since I'm relatively new to Ruby, so just left some pointer comments to myself for when I have time (hopefully later this week) to research/review in more depth... just ignore those for now.

**Comment by milind009** [comment]
@jeffwidman gentle reminder on this PR

**Comment by jeffwidman** [comment]
Sorry for the delay, I'll try to get this reviewed in detail this week.

**Comment by AlekhyaYalla** [comment]
A gentle ping on  this PR @jeffwidman

**Comment by deivid-rodriguez** [comment]
Hei! Before I review this, may I ask for an rebase + squash? It seems that the PR is out of date with latest main, and also there's a lot of commits amending previous changes that I think can all be squash to a single commit? Thank you!

**Comment by AlekhyaYalla** [comment]
@deivid-rodriguez I have done the part of rebasing. Could you please review this now?

**Comment by deivid-rodriguez** [comment]
Yes, will review this ASAP. So sorry for the delay.

File: npm_and_yarn/lib/dependabot/npm_and_yarn/update_checker/registry_finder.rb [hunk 4]
```diff
@@ -203,19 +208,20 @@ def unique_registries(registries)
...
         # rubocop:disable Metrics/PerceivedComplexity
+        # rubocop:disable Metrics/AbcSize
         def global_registry
           return @global_registry if defined? @global_registry
 
           npmrc_file&.content.to_s.scan(NPM_GLOBAL_REGISTRY_REGEX) do
             next if Regexp.last_match[:registry].include?("${")
 
-            return @global_registry = Regexp.last_match[:registry].strip
+            return @global_registry = Regexp.last_match[:registry].strip.gsub(/\s+/, "%20")
```
Comment (deivid-rodriguez) [inline npm_and_yarn/lib/dependabot/npm_and_yarn/update_checker/registry_finder.rb:218]: I think this change is unnecessary and it just a leftover from fixing merge conflicts to get the PR up to date?

File: npm_and_yarn/lib/dependabot/npm_and_yarn/update_checker/registry_finder.rb [hunk 5]
```diff
@@ -203,19 +208,20 @@ def unique_registries(registries)
...
           end
 
           yarnrc_file&.content.to_s.scan(YARN_GLOBAL_REGISTRY_REGEX) do
             next if Regexp.last_match[:registry].include?("${")
 
-            return @global_registry = Regexp.last_match[:registry].strip
+            return @global_registry = Regexp.last_match[:registry].strip.gsub(/\s+/, "%20")
```
Comment (deivid-rodriguez) [inline npm_and_yarn/lib/dependabot/npm_and_yarn/update_checker/registry_finder.rb:224]: I think this one is unnecessary too.

File: npm_and_yarn/lib/dependabot/npm_and_yarn/update_checker/registry_finder.rb [hunk 6]
```diff
@@ -232,6 +238,7 @@ def global_registry
           "https://registry.npmjs.org"
         end
         # rubocop:enable Metrics/PerceivedComplexity
+        # rubocop:enable Metrics/AbcSize
```
Comment (deivid-rodriguez) [inline npm_and_yarn/lib/dependabot/npm_and_yarn/update_checker/registry_finder.rb:241]: If the above comments are correct, this method does not need to change, so you need to undo the RuboCop skips as well.

File: npm_and_yarn/spec/fixtures/projects/npm6/npmrc_auth_token_with_space/package.json (hunk 3 above)
Comment (deivid-rodriguez) [reply npm_and_yarn/spec/fixtures/projects/npm6/npmrc_auth_token_with_space/package.json:7]: It sounds like that should be a double slash and that's why GitHub is showing it in red?

File: npm_and_yarn/spec/fixtures/projects/yarn/yarnrc_global_registry_with_space/package.json [hunk 7]
```diff
@@ -0,0 +1,25 @@
+{
+    "name": "{{ name }}",
+    "version": "1.0.0",
+    "description": "",
+    "main": "index.js",
+    "scripts": {
+        "test": "echo \"Error: no\\ test\ specified\" && exit 1",
```
Comment (deivid-rodriguez) [inline npm_and_yarn/spec/fixtures/projects/yarn/yarnrc_global_registry_with_space/package.json:7]: Same with the missing double slash. I double checked and that makes the script unrunnable, you get an error like this:

```
$ yarn test
error An unexpected error occurred: "/path/to/package.json: Bad escaped character in JSON at position 121".
```

No big deal but best to use sane fixtures in our specs!

**Review by deivid-rodriguez — COMMENTED** [review]
Just some minor comments, but the fix looks straightforward and should be ready to merge once those small things are tweaked!

Thanks so much for fixing this ❤️.

File: npm_and_yarn/spec/fixtures/projects/yarn/yarnrc_global_registry_with_space/package.json (hunk 7 above)
Comment (AlekhyaYalla) [reply npm_and_yarn/spec/fixtures/projects/yarn/yarnrc_global_registry_with_space/package.json:7]: Yup addressed them. Thanks for review @deivid-rodriguez

File: npm_and_yarn/lib/dependabot/npm_and_yarn/update_checker/registry_finder.rb (hunk 4 above)
Comment (AlekhyaYalla) [reply npm_and_yarn/lib/dependabot/npm_and_yarn/update_checker/registry_finder.rb:218]: I reason why I added this is, for fixing any spaces in global uri

**Review by deivid-rodriguez — APPROVED** [review]
Thanks, it's looking great now, I'll merge after GitHub Universe's code freeze!

**Comment by deivid-rodriguez** [comment]
Thanks so much for fixing this, and for your patience @AlekhyaYalla!

**Comment by jeffwidman** [comment]
My apologies for the slowness @AlekhyaYalla and @milind009, and many thanks to @deivid-rodriguez for helping get this across the line.

**Pull request merged**